### PR TITLE
Add Series.product

### PR DESCRIPTION
**Pull request by itholic**
Implemented `Series.product` (`Series.prod` as an alias).

*Note that since It is implemented by using `exp(sum(log(...)))` trick, it only works for positive numbers.

```python
>>> ks.Series([]).prod()
1.0
>>> ks.Series([1, 2, 3, 4, 5]).prod()
120
>>> ks.Series([1, np.nan, 3, np.nan, 5]).prod()
15.0
```

### CODE REVIEW

File: databricks/koalas/series.py [hunk 1]
```diff
@@ -4962,6 +4962,72 @@ def tail(self, n=5):
...
+        """
+        data_type = self.spark.data_type
+        if isinstance(data_type, StringType):
+            raise TypeError("can't multiply sequence by non-int of type 'str'")
+        # When number of valid values is fewer than `min_count`, pandas returns np.nan
+        if (min_count > 0) and (len(self.dropna()) < min_count):
+            return np.nan
+        # When Series is empty, pandas returns 1.0
+        if self.empty:
+            return 1.0
```
Comment (ueshin) [inline databricks/koalas/series.py:5016]: I guess we can check this later?

```py
result = spark_frame.head(1)
if len(result) == 0:
    return 1.0
elif ...
```

File: databricks/koalas/tests/test_series.py [hunk 2]
```diff
@@ -1894,3 +1895,49 @@ def test_tail(self):
...
+
+    def test_product(self):
+        pser = pd.Series([10, 20, 30, 40, 50])
+        kser = ks.from_pandas(pser)
+        self.assert_eq(pser.prod(), kser.prod())
+
+        # Containing NA values
+        pser = pd.Series([10, np.nan, 30, np.nan, 50])
+        kser = ks.from_pandas(pser)
+        self.assert_eq(isclose(pser.prod(), kser.prod()), True)
```
Comment (ueshin) [inline databricks/koalas/tests/test_series.py:1907]: Shall we use `self.assertTrue()`?

File: databricks/koalas/tests/test_series.py [hunk 3]
```diff
@@ -1894,3 +1895,49 @@ def test_tail(self):
...
+        # With `min_count` parameter
+        pser = pd.Series([10, 20, 30, 40, 50])
+        kser = ks.from_pandas(pser)
+        self.assert_eq(pser.prod(min_count=5), kser.prod(min_count=5))
+        # Using `repr` since the result of below will be `np.nan`.
+        self.assert_eq(repr(pser.prod(min_count=6)), repr(kser.prod(min_count=6)))
+
+        pser = pd.Series([10, np.nan, 30, np.nan, 50])
+        kser = ks.from_pandas(pser)
+        self.assert_eq(isclose(pser.prod(min_count=3), kser.prod(min_count=3)), True)
```
Comment (ueshin) [inline databricks/koalas/tests/test_series.py:1928]: ditto.

File: databricks/koalas/tests/test_series.py (hunk 2 above)
Comment (ueshin) [reply databricks/koalas/tests/test_series.py:1907]: btw, doesn't `self.assert_eq(pser.prod(), kser.prod(), almost=True)` work?

Comment (itholic) [reply databricks/koalas/tests/test_series.py:1907]: Oh, It works. Thanks! 👍

File: databricks/koalas/series.py (hunk 1 above)
Comment (itholic) [reply databricks/koalas/series.py:5016]: Seems better. Thanks! :)

File: databricks/koalas/series.py [hunk 4]
```diff
@@ -4962,6 +4962,72 @@ def tail(self, n=5):
...
+        1.0
+
+        This can be controlled with the ``min_count`` parameter
+
+        >>> ks.Series([]).prod(min_count=1)
+        nan
+        """
+        data_type = self.spark.data_type
+        if isinstance(data_type, StringType):
+            raise TypeError("can't multiply sequence by non-int of type 'str'")
```
Comment (ueshin) [inline databricks/koalas/series.py:5010]: What if the other data types, such as `BooleanType`, `DateType`, or `TimestampType`?

Comment (itholic) [reply databricks/koalas/series.py:5010]: I'll address it. Thanks!

File: databricks/koalas/series.py [hunk 5]
```diff
@@ -4962,6 +4965,76 @@ def tail(self, n=5):
...
         return first_series(self.to_frame().tail(n=n))
 
+    def product(self, min_count=0):
+        """
+        Return the product of the values.
+
+        .. note:: unlike pandas', Koalas' emulates product by ``exp(sum(log(...)))``
+            trick. Therefore, it only works for positive numbers.
+
+        .. note:: the result can be slightly difference from pandas since the rounding error.
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py:4975]: I think it's okay to don't mention this. Fractions can be not exactly matched anyway at Python in general.

File: databricks/koalas/series.py [hunk 6]
```diff
@@ -4962,6 +4965,76 @@ def tail(self, n=5):
...
+        >>> ks.Series([]).prod(min_count=1)
+        nan
+        """
+        # When number of valid values is fewer than `min_count`, pandas returns np.nan
+        if (min_count > 0) and (len(self.dropna()) < min_count):
+            return np.nan
+
+        data_type = self.spark.data_type
+        if isinstance(data_type, BooleanType):
+            return self.sum() == len(self)
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py:5017]: I think you can use `self.all()` instead.

File: databricks/koalas/series.py [hunk 7]
```diff
@@ -4962,6 +4965,76 @@ def tail(self, n=5):
...
+        elif isinstance(data_type, (IntegralType, FractionalType)):
+            spark_frame = self._internal.spark_frame
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py:5019]: I think you should use `resolved_copy. spark_frame` so other columns and operations are all applied before creating a Spark DataFrame.

**Comment by itholic** [comment]
Thanks, @HyukjinKwon .

File: databricks/koalas/series.py [hunk 8]
```diff
@@ -4962,6 +4965,61 @@ def tail(self, n=5):
...
+        nan
+        """
+        # When number of valid values is fewer than `min_count`, pandas returns np.nan
+        if (min_count > 0) and (len(self.dropna()) < min_count):
+            return np.nan
+
+        data_type = self.spark.data_type
+        if isinstance(data_type, BooleanType):
+            return self.all()
+        elif isinstance(data_type, (IntegralType, FractionalType)):
```
Comment (ueshin) [inline databricks/koalas/series.py:5003]: nit: `NumericType`?

File: databricks/koalas/series.py [hunk 9]
```diff
@@ -4962,6 +4965,61 @@ def tail(self, n=5):
...
+            spark_column = self.spark.column
+
+            cond = F.when(spark_column.isNull(), F.lit(1)).otherwise(spark_column)
+            spark_frame = spark_frame.select(F.exp(F.sum(F.log(cond))))
+
+            result = spark_frame.head(1)[0][0]
+            if result is None:
+                # When Series is empty, pandas returns 1.0
+                return 1.0
+            elif isinstance(data_type, (LongType, BooleanType)):
```
Comment (ueshin) [inline databricks/koalas/series.py:5014]: nit: `BooleanType` is not needed since it's handled above.

Comment (ueshin) [reply databricks/koalas/series.py:5014]: btw, `IntegralType` instead of `LongType`?

File: databricks/koalas/series.py [hunk 10]
```diff
@@ -44,6 +44,9 @@
     NumericType,
     StringType,
     StructType,
+    TimestampType,
```
Comment (ueshin) [inline databricks/koalas/series.py:47]: nit: `TimestampType` is not used?

**Comment by HyukjinKwon** [comment]
@itholic can you address the comments above?

File: databricks/koalas/series.py (hunk 10 above)
Comment (itholic) [reply databricks/koalas/series.py]: Thanks, @ueshin .

**Pull request merged**